staging/src/k8s.io/apiserver/pkg/util/webhook/client.go

```diff
@@ -14,32 +14,40 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package config
+package webhook
 
 import (
 	"context"
-	"encoding/json"
 	"errors"
 	"fmt"
 	"net"
 	"net/url"
 
-	lru "github.com/hashicorp/golang-lru"
-	admissionv1beta1 "k8s.io/api/admission/v1beta1"
-	"k8s.io/api/admissionregistration/v1beta1"
+	"github.com/hashicorp/golang-lru"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/apimachinery/pkg/runtime/serializer"
 	utilerrors "k8s.io/apimachinery/pkg/util/errors"
-	webhookerrors "k8s.io/apiserver/pkg/admission/plugin/webhook/errors"
 	"k8s.io/client-go/rest"
 )
 
 const (
 	defaultCacheSize = 200
 )
 
+// ClientConfigInterface is the interface to access webhook client config to share webhook connect code between
+// admission webhooks and CRD conversion webhooks.
+type ClientConfigInterface interface {
+	GetURL() *string
+	GetCABundle() []byte
+	GetServiceName() *string
```

Comment:
why all these string pointers? odd.

Reply:
Because Service is optional.

Reply:
We usually use empty strings for unset strings.

Reply:
Can any of those be a valid empty string? like a valid Path or Service Namespace?

Reply:
Actually we do not use empty string everywhere for that (and I think API convention is to user pointer for optional strings we just don't enforced that well). Example of that is Service.Path in this same struct that is a string pointer.